### PR TITLE
#165010601 Homepage

### PR DESCRIPTION
**Pull request by Andhrah**
#### What does this PR do?
It enables users to report an article

#### Description of Task to be completed?
- write test
- create reusable navbar component
- create reusable footer component
- create homepage component
- style the components
- create an action creator for getting recently publish articles
- create an action creator for getting article categories
- create reusable article reducer
- create reusable article categories reducer
- create reusable article component
- create reusable authors component
- create reusable category component
- install babel plugin and set it up for using async-await in redux-thunk middleware
- build out homepage

#### How should this be manually tested?
- Follow the steps below:
  - Run the following commands in your terminal
`git clone https://github.com/andela/apollo-ah-frontend.git`
`cd apollo-ah-frontend`
- Run `npm install` to install all dependencies.
- Start the app 
N/A
#### What are the relevant pivotal tracker stories?
[#165010601](https://www.pivotaltracker.com/story/show/165010601)
#### Screenshots (if appropriate)
N/A

#### Questions:

### CODE REVIEW

File: src/actions/articleAction.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  type: actionTypes,
+  payload: articles
+});
+
+export const getArticles = () => async(dispatch) => {
+  try {
+    const response = await http.get('/articles?size=12');
+    const { data: { data } } = response;
+    if (response.status === 404) {
+      dispatch(articleActionResponse(GET_ARTICLES_FAILURE, 'Article not found'));
```
Comment (aanchirinah) [inline src/actions/articleAction.js]: #### Nitpick

- Add Asynchronous caveat ie. `LOADING`

File: src/actions/articleAction.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import { GET_ARTICLES_SUCCESS, GET_ARTICLES_FAILURE } from './actionTypes';
```
Comment (aanchirinah) [inline src/actions/articleAction.js]: #### Nitpick

- Use actionType generator (selectors)

File: src/actions/articleCategoryAction.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import { GET_ARTICLES_CATEGORY_SUCCESS, GET_ARTICLES_CATEGORY_FAILURE } from './actionTypes';
+import http from '../services/http';
+
+export const categoryActionResponse = (actionType, category) => ({
+  type: actionType,
+  payload: category
+});
+
+export const getArticlesCategory = () => async(dispatch) => {
+  try {
```
Comment (aanchirinah) [inline src/actions/articleCategoryAction.js:26]: #### Nitpick

- Add loading action type

**Review by aanchirinah — COMMENTED** [review]
Awesome job, kindly attend to these comments

File: src/actions/actionTypes.js [hunk 4]
```diff
@@ -0,0 +1,11 @@
+import keyMirror from 'keymirror';
+
+export const ACTIONS = keyMirror({
```
Comment (aanchirinah) [inline src/actions/actionTypes.js]: #### Critical
- Not the kinda selector generator I was looking for, it's not supposed to exceed more than 5 lines of code.

File: src/actions/articleAction.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    const { data: { data } } = response;
+    if (response.status === 404) {
+      dispatch(articleAction(ACTIONS.GET_ARTICLES_FAILURE, 'Article not found'));
+      dispatch(articleAction(ACTIONS.STOP_LOADING, 'stopped'));
+    }
+    dispatch(articleAction(ACTIONS.GET_ARTICLES_SUCCESS, data));
+    dispatch(articleAction(ACTIONS.STOP_LOADING, 'stopped'));
+  } catch (err) {
+    dispatch(articleAction(ACTIONS.STOP_LOADING, 'stopped'));
+    return err;
```
Comment (aanchirinah) [inline src/actions/articleAction.js]: #### Nitpick
- Why the return statement?

File: src/actions/articleCategoryAction.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    const { data: { data } } = response;
+    if (response.status === 404) {
+      dispatch(categoryActionResponse(ACTIONS.GET_ARTICLES_CATEGORY_FAILURE, 'Category not found'));
+      categoryActionResponse(ACTIONS.STOP_LOADING, 'stopped');
+    }
+    dispatch(categoryActionResponse(ACTIONS.GET_ARTICLES_CATEGORY_SUCCESS, data));
+    categoryActionResponse(ACTIONS.STOP_LOADING, 'stopped');
+  } catch (err) {
+    categoryActionResponse(ACTIONS.STOP_LOADING, 'stopped');
+    return err;
```
Comment (aanchirinah) [inline src/actions/articleCategoryAction.js]: #### Nitpick
- Please lets not return errors directly, let's use the dispatched action.

File: src/components/HomePage.jsx [hunk 7]
```diff
@@ -1,19 +1,225 @@
...
+    const { getArticles, getArticlesCategory } = this.props;
+    await Promise.all([getArticles(), getArticlesCategory()]);
+  }
+
+  /**
+   * @description Gets the averahe ratings of author's article
+   * @param {object} article - each authors article
+   * @return {array} fiveStaAauhtors - author's rated highly
+   */
+  averageRatings = (article) => {
```
Comment (aanchirinah) [inline src/components/HomePage.jsx:34]: #### Nitpick
- Resolve this if you have already removed this out of here

File: src/components/HomePage.jsx [hunk 8]
```diff
@@ -1,19 +1,225 @@
...
+      }
+    });
+  }
+
+  /**
+ * @description set the state with start authors
+ * @return {void}
+ */
+  setAuthorsState = (rating) => (this.setState({
+    fiveStarAuthors: [...this.state.fiveStarAuthors, rating],
```
Comment (aanchirinah) [inline src/components/HomePage.jsx:44]: #### Comment
- Can you explain why you are using the spread operator here?

File: src/components/HomePage.jsx [hunk 9]
```diff
@@ -1,19 +1,225 @@
...
+        <Navbar
+          categories={'Categories'}
+          authors={'Authors'}
+          bookmark={'Bookmarks'}
+          search={'Search'}
+          write={'Write an areticle'}
+          login={'Login'}
+        />
+
+        <main className="main-body">
```
Comment (aanchirinah) [inline src/components/HomePage.jsx]: #### Nitpick
- Resolve this if you have already created the view file for this.

File: src/views/Articles.jsx [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+ * @param {props} articles - all authors article
+ * @return {JSX} 
+ */
+class Articles extends Component {
+  /**
+ * @description time an article was created
+ * @param {object} article - each authors article
+ * @return {array} fiveStaAauhtors - author's rated highly
+ */
+  time(article) {
```
Comment (aanchirinah) [inline src/views/Articles.jsx]: #### Nitpick
- Should be isolated out of here.

File: src/views/Navbar.jsx [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+              src="#"
+            />
+          </button>
+          <div
+            className="dropdown-menu dropdown-menu-right"
+            aria-labelledby="dropdownMenuButton"
+          >
+            <Link className="dropdown-item" to="#">
+              Write an articlLink
+        </Link>
```
Comment (aanchirinah) [inline src/views/Navbar.jsx]: #### Critical
- Against DRY principle.

File: src/views/placeholderLoader.jsx [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+        <ContentLoader 
+          height={230}
+          width={300}
+          speed={3}
+          primaryColor="#f3f3f3"
+          secondaryColor="rgba(217, 196, 224, 0.32)"
+        />
+      </div>
+      
+      <div className="col-md-4 col-sm-6" style={{marginBottom: 30}}>
```
Comment (aanchirinah) [inline src/views/placeholderLoader.jsx]: #### Critical
- Against DRY principle.

**Review by aanchirinah — CHANGES_REQUESTED** [review]
Good job, can you kindly attend to my comments and requested changes.

File: src/actions/articleAction.js (hunk 5 above)
Comment (Andhrah) [reply src/actions/articleAction.js]: I'll attend to that now

File: src/actions/articleCategoryAction.js (hunk 6 above)
Comment (Andhrah) [reply src/actions/articleCategoryAction.js]: okay thanks

File: src/actions/articleCategoryAction.js (hunk 3 above)
Comment (Andhrah) [reply src/actions/articleCategoryAction.js:26]: Done

File: src/actions/articleAction.js (hunk 1 above)
Comment (Andhrah) [reply src/actions/articleAction.js]: Done, thank you

File: src/components/HomePage.jsx (hunk 9 above)
Comment (Andhrah) [reply src/components/HomePage.jsx]: Alright

**Comment by aanchirinah** [comment]
@Andraquin What's the status of this now?

File: src/components/HomePage.jsx (hunk 8 above)
Comment (Andhrah) [reply src/components/HomePage.jsx:44]: Yeah, I'm using it to update my `fiveStarAuthors` state

File: src/views/Articles.jsx (hunk 10 above)
Comment (Andhrah) [reply src/views/Articles.jsx]: Done

**Comment by Andhrah** [comment]
> Awesome job, kindly attend to these comments

I have attended to your comments

**Comment by Andhrah** [comment]
> @Andraquin What's the status of this now?

It's ready now

File: src/components/HomePage.jsx [hunk 13]
```diff
@@ -1,19 +1,121 @@
...
+  /**
+ * @description set the state with start authors
+ * @return {void}
+ */
+  setAuthorsState = (rating) => (this.setState({
+    fiveStarAuthors: [...this.state.fiveStarAuthors, rating],
+  }));
+
+  /**
+ * @description found uthors
```
Comment (aanchirinah) [inline src/components/HomePage.jsx]: #### Nitpick
- Did you mean this rather
```suggestion
 * @description found authors
```

**Review by aanchirinah — APPROVED** [review]
Approved but you have typos. @Andraquin

**Review by fejiroofficial — APPROVED** [review]
Looks good. Great Job

**Comment by fejiroofficial** [comment]
It seems code climate is not your friend. Great Job altogether

**Pull request merged**